### PR TITLE
semaphore added

### PR DESCRIPTION
**Pull request by gsbelarus**

### CODE REVIEW

File: src/Semaphore.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      this._queue.shift()?.();
+    } else {
+      this._permits += 1;
+    }
+
+    if (this._permits > this._maxAllowed) {
+      console.error('Release() has been called without previous acquire!');
+    }
+  }
+}
```
Comment (gsbelarus) [inline src/Semaphore.ts]: Overall, the code patch looks good. Here are some observations and suggestions:

1. It's a TypeScript code patch, indicating that you're using a static type-checking version of JavaScript.

2. The `Semaphore` class represents a semaphore with an optional maximum allowed permits. The class provides methods for acquiring and releasing permits.

3. The class has private properties `_maxAllowed`, `_permits`, and `_queue`.

4. Consider making the private properties `readonly` to prevent accidental modifications.

5. The constructor initializes the `_maxAllowed` and `_permits` properties based on the provided or default values.

6. The class has getter methods for `permits`, `queueLength`, and `acquired` properties for retrieving their current values.

7. The `acquire` method is an asynchronous function that allows acquiring a permit. If there are available permits, it reduces the number of permits by one and returns immediately. Otherwise, it adds a resolver function to the `_queue` array and returns a promise that will be resolved when a permit becomes available.

8. The `release` method releases a permit by either invoking the next resolver function in the `_queue` array if there are waiting promises, or incrementing the number of permits.

9. There's an error check in the `release` method to log a message if `release` is called without a previous `acquire`. However, it would be better to throw an error instead of logging to console.error.

10. Overall, the code follows a simple and straightforward approach for implementing a semaphore.

Improvement suggestions:

- Considering adding safety checks to ensure that the number of acquired and released permits does not exceed the specified maximum.

- If performance is a concern, optimizing the queue implementation could be beneficial, such as using a linked list instead of an array for the `_queue` property.

- Adding unit tests to cover different scenarios and edge cases would be valuable to ensure the functionality and correctness of the semaphore class.

- Documentation comments describing the purpose and behavior of the class and its methods would enhance code readability and maintainability.

Remember to adapt these suggestions based on your specific requirements and constraints.
